### ISSUE TITLE
Handle errors when sending tokens within Aurora to NEAR 

### ISSUE THREAD

**Issue by mfornet** [aurora-is-near/aurora-engine#269]
Currently the way exit precompiles work, is that it:
1. verifies the amount of tokens is owned by the user inside Aurora
2. burns those tokens
3. schedule a cross contract call to mint/create withdrawal event on NEAR

Currently if step 3) fails for some reason, tokens are burnt within aurora, and not minted on NEAR, so basically they are lost. We need to handle failures on step 3. This is, we schedule yet another call after the exit call where we see if the tx succeeded or not, in case it didn't succeed funds should be refunded.

This was first detected on [this tx](https://explorer.near.org/transactions/BHKYWbAzxXXAhCsNbageaG9dcYfZ3kjRvW9YPdmoRr42). Where `ft_transfer` failed because receiver account was not registered in the NEP-141 tokens. Currently this tokens are being held by `aurora` account, but they were not refunded to the user.

**Comment by joshuajbouw** [comment]
That is rather unfortunate. We will look into this.

**Comment by sept-en** [comment]
Update on this: the tokens were manually refunded to the user on [this tx](https://explorer.mainnet.near.org/transactions/27UQ4Mgivs5y6z9BafPs55nB45bgHsxYdf72LnMtPUii)

**Comment by mrLSD** [comment]
> This is, we schedule yet another call after the exit call where we see if the tx succeeded or not, in case it didn't succeed funds should be refunded.

Seems we can't do that. The main reason promises are async. Also, we can't analyze TX inside `aurora` contract. We should have some mechanism outside `aurora` contract that will check exit precompiles outcome and fire new call for a refund if something wrong. 

Obvious solutions:
1. less reliable but simple - we can do it via front-end action
2. We can implement and run it as some service that can analyze exit precompiles incidents and fire-related actions to solve it.

@mfornet, @joshuajbouw let's choose the best solution for that.

**Comment by birchmd** [comment]
@mrLSD I think you should be able to attach a callback to the promise (using the `then` API). That callback should check for success of the previous call and revert the transfer to the precompile

**Comment by mrLSD** [comment]
I investigated that, and still have questions.

So, the `callback` function for failing case for that promise looks really dangerous. It seems possible bugs, which are difficult to catch.

Anyway, we have `promise` schedule:
https://github.com/aurora-is-near/aurora-engine/blob/3eace8c03050252f8df0cf7709cda16a8e019ee7/src/precompiles/native.rs#L167

So I'll implement a dependent `promise` burn from `aurora` account and mint to the `receiver` account.

But do we really clear that the approach doesn't have pitfalls?
@sept-en @joshuajbouw

**Comment by birchmd** [comment]
> So I'll implement a dependent promise burn from aurora account and mint to the receiver account.

No, this is not what you should do. The transfer that needs to be reverted in the the engine itself. You need to transfer the tokens back from the precompile address to the user's address in the engine. The NEP-141 balances do not need to be changed because this flow only triggers when the NEP-141 transaction was unsuccessful.

> So, the callback function for failing case for that promise looks really dangerous. It seems possible bugs, which are difficult to catch.

I'm not sure what you are worried about. Could you elaborate more about the danger here?

**Comment by birchmd** [comment]
@mfornet Could you take a look at https://github.com/aurora-is-near/aurora-engine/pull/284#discussion_r715895103 ? Have you thought about the case where the exit precompile was invoked as part of a larger transaction?

**Comment by mfornet** [comment]
@birchmd answered on this comment: https://github.com/aurora-is-near/aurora-engine/pull/284#discussion_r718391734

**Comment by birchmd** [comment]
@mfornet , I've replied https://github.com/aurora-is-near/aurora-engine/pull/284#discussion_r718467922
